vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java

```diff
@@ -400,6 +408,395 @@ public ElementQuery<T> withoutTheme(String theme) {
+     * For matching a placeholder exactly, see {@link #withPlaceholder(String)},
+     * and for matching a placeholder partially, see {@link #withPlaceholderContaining(String)}.
+     * <p>
+     * This method can be used for other placeholder matching needs,
+     * such as performing a case-insensitive match:
+     * {@code withPlaceholder("name", String::equalsIgnoreCase)}
+     * <br> a placeholder-ending match:
+     * {@code withPlaceholder(" Name", String::endsWith)}
+     * <br> or a regular expression match:
+     * {@code withPlaceholder("(First|Last) Name", String::matches)}
+     *
+     * @param text
+     *          the text to compare with the placeholder; not null
+     * @param comparison
+     *          the comparison to use when comparing the placeholder
+     *          against the supplied text
+     *          (i.e., {@code comparison.test(elementPlaceholder, text)}; not null
+     * @return this element query instance for chaining
+     *
+     * @see #withPlaceholder(String)
+     * @see #withPlaceholderContaining(String)
+     */
+    public ElementQuery<T> withPlaceholder(String text, BiPredicate<String, String> comparison) {
+        Objects.requireNonNull(text, NULL_TEXT_MSG);
+        Objects.requireNonNull(comparison, NULL_COMPARISON_MSG);
+        return withCondition(element -> (element instanceof HasPlaceholder hasPlaceholder) &&
+                comparison.test(hasPlaceholder.getPlaceholder(), text));
+    }
+
+    /**
+     * Requires the element's placeholder to exactly match the given placeholder value.
+     * <p>
+     * For partially matching text within the placeholder, see
+     * {@link #withPlaceholderContaining(String)}.
+     *
+     * @param placeholder
+     *            the placeholder to match
+     * @return this element query instance for chaining
+     *
+     * @see #withPlaceholderContaining(String)
+     * @see #withPlaceholder(String, BiPredicate)
+     */
+    public ElementQuery<T> withPlaceholder(String placeholder) {
+        return withPlaceholder(placeholder, String::equals);
+    }
+
+    /**
+     * Requires the element's placeholder to partially match the given text value.
+     * <p>
+     * For exactly matching the placeholder, see
+     * {@link #withPlaceholder(String)}.
+     *
+     * @param text
+     *          the text to match
+     * @return this element query instance for chaining
+     *
+     * @see #withPlaceholder(String)
+     * @see #withPlaceholder(String, BiPredicate)
+     */
+    public ElementQuery<T> withPlaceholderContaining(String text) {
+        return withPlaceholder(text, String::contains);
+    }
+
+    /**
+     * Requires the element's caption (i.e., label, placeholder, or text label)
+     * to satisfy the given comparison with the supplied text.
+     * <p>
+     * If the element has a label, the label value is used.
+     * If the element can have a label, but its value is empty and it has a placeholder,
+     * the placeholder value is used.
```

Comment:
This statement is a bit confusing to me: it looks like that `placeholder` is used only if the element has an empty label, not if it does not support a label.